### ISSUE TITLE
How to use estimate_threshold to modify --mod-binary-threshold 

### ISSUE THREAD

**Issue by jeffsteimle** [nanoporetech/megalodon#251]
I recently used megalodon to call 5hmC and 5mC on mESC that are either wildtype, completely lacking the TET enzymes (no 5hmC) and completely lacking the DNMT (no 5mC or 5hmC) as a control experiment. For 5mC, everything looks what I would expect within some reasonable error (~1% of sites are being called 5mC in the DNMT mutants), which I can live with for now. However, for 5hmC, the TET and DNMT mutants are both calling around 5-6% 5hmC while WT is calling closer to 10%. Based on this, I would simplistically assume 1 out of 2 calls are false positives (!!!). 

I saw some discussion on here about adjusting the --mod-binary-threshold and using the estimate_threshold tool in megalodon_extras.

For example, below are the outputs I got for 5hmC (top 2) and 5mC (bottom 2) for the WT and TET mutant. 

**How would I utilize these values to adjust the parameters of megalodon assuming the ground truth that the TET (and DNMT) mutants should have 0% 5hmC?**

### WT - 5hmC
Fraction mod: 0.6085456475583864
Threshold: -0.9037592721696484

### TET Mutant - 5hmC
Fraction mod: 0.39360006710283507
Threshold: -1.9788528778814276

### WT - 5mC
Fraction mod: 0.8402296360485269
Threshold: 4.68547797827339

### TET Mutant - 5mC
Fraction mod: 0.8188119003031064
Threshold: 5.45577697836797

---

Sorry if this was posted elsewhere! I couldn't find anything specific.

**Comment by marcus1487** [comment]
A 50% false positive rate would absolutely be quite problematic. Our internal PCR controls showed false positive rates much closer to 1-2% (as shown in my presentation at NCM). Though this is a more research focused model, so we have not tested it against a wide range of organisms. 

For the results from the `megalodon_extras modified_bases estimate_threshold` command, I believe that this is not the intended use for this command. This command is intended to estimate the `--ref-mod-threshold` parameter and not the `--mod-binary-threshold` parameter. This command was intended to assist in marking up training dataset with randomly incorporated modified bases. The results on such native samples described here does not have a useful interpretation. Have a look at the [documentation for this command here](https://nanoporetech.github.io/megalodon/extras_modified_bases.html#megalodon-extras-modified-bases-estimate-threshold).

**Comment by jeffsteimle** [comment]
@marcus1487 Thank you so much for getting back to me quickly. 

I see now that `megalodon_extra modified_bases estimate_threshold` isn't used for what I was thinking it was used for.

Clearly the default parameters worked well for 5mC, and actually very similar to what you find. Let me ask the question another way, **is there a best practices method for adjusting the parameters to try and fit the data? Is this more of a guess and check approach?**

**Comment by jeffsteimle** [comment]
Out of curiosity, I tried adjusting 	`--mod-binary-threshold` incrementally on a consistent subset of data from each of WT, TET mutant, and DNMT mutant samples. I tried adjusting the value from the default (0.80) to 0.75, 0.85, 0.90, 0.95, 0.99, and 0.999. 

My observations:
1. The total alignment rate is severely hurt by increasing this parameter above 0.95, as we might expect as the threshold approaches 1. At 0.95 and below, there is an effect, but it is much more incremental and its pros/cons discussion then. Clearly 0.999 is too strict and calls in general fall apart completely.
2. 0.99 is very strict and my total number of CG sites is less than half 0.95. However, for DNMT mutants, the 5mC and 5hmC rates were similar, 0.7% and 0.8% respectively. For TET mutants, only 0.7% 5hmC. This threshold also saw a jump in the 5mC calls for WT and TET mutants as compared to 0.95. This appears to be a result of the total C's dropping out faster than the 5mC.
3. The percent of 5mC/C calls were largely unaffected across all samples by adjusting this parameter from 0.75-0.95. The models must very clearly distinguish 5mC from canonical C. **Great to see!**
4. There was a beautiful relationship between `--mod-binary-threshold` and the percent of 5hmC to C calls. Every 5% adjustment to the threshold, reduced the calls by 1% for the TET mutants (5.5%, 4.8%, 3.5%, 2.6%, 1.6%, 0.7%). Nice when math works that way.
5. Unless there is something I am not seeing or understanding, which is very possible, using a non-default `--mod-binary-threshold 0.95` may be a great way to improve confidence of 5hmC calls with little to no effect on 5mC and only a slight loss of total reads. For reference, I see around 1.5% and 1.6% 5hmC for DNMT and TET mutant respectively, and 6.8% in WT mESCs. With assumptions, maybe 1 in 4 false positives, much improved from the 1 in 2 I saw at default. If I was only interested in 5mC, I would probably stick to the default 0.8.

The exact parameters I used below. I used the HAC models for all tests, because I found little difference, other than time to finish, between HAC and SUP models under the default threshold for these samples. 

All data sequenced on a minION mk1b (minKNOW 21.11.17) with a R10.4 flow cell, using the SQK-LSK112 prep kit.
Megalodon version: 2.4.1
Guppy: 6.0.1

	megalodon \
	Fast5/Testing \
	--outputs basecalls mappings per_read_mods mods mod_mappings \
	--output-directory megalodon/Testing \
	--sort-mappings \
	--guppy-config dna_r10.4_e8.1_hac.cfg \
	--guppy-server-path /tools/src/ont-guppy_6.0.1_linux64/ont-guppy/bin/guppy_basecall_server \
	--remora-modified-bases dna_r10.4_e8.1 hac 0.0.0 5hmc_5mc CG 0 \
	--reference /ref/Ensembl_GRCm38.94/genome.fa \
	--processes 60 \
	--devices 0 \
	--sort-mappings \
	--mod-output-formats bedmethyl wiggle \
	--guppy-params "--chunks_per_runner 112" \
	--guppy-timeout 1200 \
	--suppress-queues-status \
	--mod-binary-threshold 0.95
